### ISSUE TITLE
将CNN替换成resnet后从头训练ocr, loss下降,精度不高, 是咋回事?

### ISSUE THREAD

**Issue by wenston2006** [chineseocr/chineseocr#358]
resnet部分定义如下, 然后将network_torch.py中的cnn用下面定义的ResNet类替换掉:

from torch import nn
import torch as t
from torch.nn import functional as F

class ResidualBlock(nn.Module):
    '''
    实现子module: Residual Block
    '''

    def __init__(self, block_id, blockin_id, inchannel, outchannel, stride=1, shortcut=None):
        super(ResidualBlock, self).__init__()
        self.left = nn.Sequential()
        self.left.add_module('rb_{0}_{0}_left_Conv1'.format(block_id,blockin_id),
                nn.Conv2d(inchannel, outchannel, 3, stride, 1, bias=False))
        self.left.add_module('rb_{0}_{0}_left_BatchNorm1'.format(block_id,blockin_id),
            nn.BatchNorm2d(outchannel))
        self.left.add_module('rb_{0}_{0}_left_Relu1'.format(block_id,blockin_id),
            nn.ReLU(inplace=True))
        self.left.add_module('rb_{0}_{0}_left_Conv2'.format(block_id,blockin_id),
            nn.Conv2d(outchannel, outchannel, 3, 1, 1, bias=False))
        self.left.add_module('rb_{0}_{0}_left_BatchNorm2'.format(block_id,blockin_id),
            nn.BatchNorm2d(outchannel))
        self.right = shortcut

    def forward(self, x):
        out = self.left(x)
        residual = x if self.right is None else self.right(x)
        out += residual
        return F.relu(out)

class ResNet(nn.Module):
    '''
    实现主module：ResNet34
    ResNet34 包含多个layer，每个layer又包含多个residual block
    用子module来实现residual block，用_make_layer函数来实现layer
    '''

    def __init__(self, num_classes=1000):
        super(ResNet, self).__init__()
        # 前几层图像转换
        self.pre = nn.Sequential()
        self.pre.add_module('pre_Conv1',
            nn.Conv2d(1, 32, (3, 3), (1, 1), (1, 1), bias=False))
        self.pre.add_module('pre_BatchNorm1',
            nn.BatchNorm2d(32))
        self.pre.add_module('pre_Relu1',
            nn.ReLU(inplace=True))
        self.pre.add_module('pre_Conv2',
            nn.Conv2d(32, 64, (3, 3), (1, 1), (1, 1), bias=False))
        self.pre.add_module('pre_BatchNorm2',
            nn.BatchNorm2d(64))
        self.pre.add_module('pre_Relu2',
            nn.ReLU(inplace=True))

        # 重复的layer，分别有3，4，6，3个residual block
        self.layer1 = self._make_layer(1, 64, 128, 1)
        self.layer2 = self._make_layer(2, 128, 256, 2)
        self.layer3 = self._make_layer(3, 256, 512, 5)
        self.layer4 = self._make_layer(4, 512, 512, 3)

        # pool layer
        self.pool1 = nn.MaxPool2d((2, 2), (2, 2), (0, 0))
        self.pool2 = nn.MaxPool2d((2, 2), (2, 2), (0, 0))
        self.pool3 = nn.MaxPool2d((2, 2), (2, 1), (0, 1))

        # post conv
        self.conv1_post = nn.Sequential()
        self.conv1_post.add_module('cp1_Conv1',
            nn.Conv2d(128, 128, (3, 3), (1, 1), (1, 1), bias=False))
        self.conv1_post.add_module('cp1_BatchNorm1',
            nn.BatchNorm2d(128))
        self.conv1_post.add_module('cp1_Relu1',
            nn.ReLU(inplace=True))

        self.conv2_post = nn.Sequential()
        self.conv2_post.add_module('cp2_Conv1',
            nn.Conv2d(256, 256, (3, 3), (1, 1), (1, 1), bias=False))
        self.conv2_post.add_module('cp2_BatchNorm1',
            nn.BatchNorm2d(256))
        self.conv2_post.add_module('cp2_Relu1',
            nn.ReLU(inplace=True))

        self.conv3_post = nn.Sequential()
        self.conv3_post.add_module('cp3_Conv1',
            nn.Conv2d(512, 512, (3, 3), (1, 1), (1, 1), bias=False))
        self.conv3_post.add_module('cp3_BatchNorm1',
            nn.BatchNorm2d(512))
        self.conv3_post.add_module('cp3_Relu1',
            nn.ReLU(inplace=True))

        self.conv4_post = nn.Sequential()
        self.conv4_post.add_module('cp4_Conv1',
            nn.Conv2d(512, 512, (2, 2), (2, 1), (0, 1), bias=False))
        self.conv4_post.add_module('cp4_BatchNorm1',
            nn.BatchNorm2d(512))
        self.conv4_post.add_module('cp4_Relu1',
            nn.ReLU(inplace=True))
        self.conv4_post.add_module('cp4_Conv2',
            nn.Conv2d(512, 512, (2, 2), (1, 1), (0, 0), bias=False))
        self.conv4_post.add_module('cp4_BatchNorm2',
            nn.BatchNorm2d(512))
        self.conv4_post.add_module('cp4_Relu2',
            nn.ReLU(inplace=True))

    def _make_layer(self, block_id, inchannel, outchannel, block_num, stride=1):
        '''
        构建layer,包含多个residual block
        '''
        shortcut = nn.Sequential()
        shortcut.add_module('rb_{0}_sc_Conv1'.format(block_id),
            nn.Conv2d(inchannel, outchannel, 1, stride, bias=False))
        shortcut.add_module('rb_{0}_sc_BatchNorm'.format(block_id),
            nn.BatchNorm2d(outchannel))

        layers = []
        layers.append(ResidualBlock(block_id, 0, inchannel, outchannel, stride, shortcut))

        for i in range(1, block_num):
            layers.append(ResidualBlock(block_id, i, outchannel, outchannel))
        return nn.Sequential(*layers)

    def forward(self, x):
        # x = self.cnn(x)

        x = self.pre(x)
        x = self.pool1(x)
        x = self.layer1(x)
        x = self.conv1_post(x)

        x = self.pool2(x)
        x = self.layer2(x)
        x = self.conv2_post(x)

        x = self.pool3(x)
        x = self.layer3(x)
        x = self.conv3_post(x)

        x = self.layer4(x)
        x = self.conv4_post(x)
        return x

**Comment by myhub** [comment]
我最近做的 http://github.com/myhub/tr 也把cnn改成resnet了，感觉效果没变好也没变差，也说不清楚为什么，不是像你说的效果反而变差

**Comment by wenlihaoyu** [comment]
关键不在模型，而在数据，当前模型已经很好，没有必要替换为其他模型，

**Comment by wenston2006** [comment]
谢谢两位的回复, 我把resnet去掉了几层, 只剩下几个简单的卷基层后, 训练收敛了, 之前的训练开头把除loss层外的其他层冻结了进行训练,我猜测跟着有关,不过我目前还没来得及验证.
